### PR TITLE
Removing AKE deps section and defining KE1/KE2/KE3 structs

### PR DESCRIPTION
**Pull request by kevinlewi**
As discussed offline, this will remove the "Authenticated Key Exchange (AKE) Protocol" subsection from the "Cryptographic Dependencies" section. This also means removing the definitions of the subroutines "AuthInit", "AuthResponse", "AuthFinish".

This is also an opportunity to clearly define the structs KE1, KE2, and KE3 in the "AKE Messages" section.

Partially addresses #336, but I think in order to complete #336, we should also re-order the sections in the Online AKE stage slightly, so that we aren't at any point referencing things that haven't yet been defined. This means moving the "Credential Retrieval" subsection up top, followed by "AKE messages", followed by "Client Authentication Functions" and "Server Authentication Functions". Not addressing this reordering in this PR, but will do so in a follow-up one.

### CODE REVIEW

File: draft-irtf-cfrg-opaque.md [hunk 1]
```diff
@@ -551,15 +516,8 @@ The authenticated key exchange flow is shown below:
...
 messages of the concurrent execution of the key recovery process (OPRF) and the
-authenticated key exchange (AKE):
-
-- `KE1` is composed of the `CredentialRequest` and `AuthInit` messages
-- `KE2` is composed of the `CredentialResponse` and `AuthResponse` messages
-- `KE3` represents the `AuthFinish` message
-
-The `CredentialRequest` and `CredentialResponse` message contents and wire
-format are specified in {{cred-retrieval}}, and those of `AuthInit`,
-`AuthResponse` and `AuthFinish` are specified in {{ake-messages}}.
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:562]: Why remove this?

File: draft-irtf-cfrg-opaque.md [hunk 2]
```diff
@@ -1272,23 +1230,29 @@ server functions, respectively.
 
 ~~~
 struct {
+  CredentialRequest credential_request;
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1233]:
```suggestion
  uint8 blinded_element[Noe];
```

File: draft-irtf-cfrg-opaque.md [hunk 3]
```diff
@@ -1272,23 +1230,29 @@ server functions, respectively.
...
   uint8 client_nonce[Nn];
   uint8 client_keyshare[Npk];
-} AuthInit;
+} KE1;
 ~~~
 
+credential_request: a CredentialRequest structure.
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1239]:
```suggestion
blinded_element: A serialized OPRF blinded element of fixed size Noe.
```

File: draft-irtf-cfrg-opaque.md [hunk 4]
```diff
@@ -1272,23 +1230,29 @@ server functions, respectively.
...
+
 client_nonce : A fresh randomly generated nonce of length Nn.
 
 client_keyshare : Client ephemeral key share of fixed size Npk.
 
 ~~~
 struct {
+  CredentialResponse credential_response;
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1247]:
```suggestion
  uint8 evaluated_message[Noe];
  uint8 masking_nonce[Nn];
  uint8 masked_response[Npk + Ne];
```

File: draft-irtf-cfrg-opaque.md [hunk 5]
```diff
@@ -1272,23 +1230,29 @@ server functions, respectively.
...
   uint8 client_keyshare[Npk];
-} AuthInit;
+} KE1;
 ~~~
 
+credential_request: a CredentialRequest structure.
+
 client_nonce : A fresh randomly generated nonce of length Nn.
 
 client_keyshare : Client ephemeral key share of fixed size Npk.
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1243]:
```suggestion
client_keyshare : Serialized client ephemeral key share of fixed size Npk.
```

File: draft-irtf-cfrg-opaque.md [hunk 6]
```diff
@@ -1272,23 +1230,29 @@ server functions, respectively.
...
 struct {
+  CredentialResponse credential_response;
   uint8 server_nonce[Nn];
   uint8 server_keyshare[Npk];
   uint8 server_mac[Nm];
-} AuthResponse;
+} KE2;
 ~~~
 
+credential_response: a CredentialResponse structure.
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1254]: evaluated_message: A serialized OPRF evaluated element of fixed size Noe.

masking_nonce: A nonce used for the confidentiality of the masked_response field of fixed size Nn.

masked_response: An encrypted form of the server's public key and the client's `Envelope` structure of fixed size Npk + Ne.

File: draft-irtf-cfrg-opaque.md [hunk 7]
```diff
@@ -1300,7 +1264,7 @@ computed using Km2, defined below.
 ~~~
 struct {
   uint8 client_mac[Nm];
-} AuthFinish;
+} KE3;
 ~~~
 
 client_mac : An authentication tag computed over the handshake transcript
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1270]:
```suggestion
client_mac: An authentication tag computed over the handshake transcript of fixed size Nm.
```

**Review by bytemare — CHANGES_REQUESTED** [review]

**Comment by chris-wood** [comment]
This change looks good, but does it match the reference implementation?

File: draft-irtf-cfrg-opaque.md (hunk 1 above)
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:562]: It doesn't seem to add any value (to me, after not having looked at it for quite some time).

**Comment by kevinlewi** [comment]
@chris-wood : These changes have no impact on the implementation, as they are purely editorial.

@bytemare : Please take a look at the latest updates here. As discussed offline, I have kept the AuthInit/AuthResponse structs (renamed AuthInit to AuthRequest though), so that KE1 and KE2 only have two elements. I moved the text about explaining the structure of KE1/KE2 to be within the AKE messages section.

**Comment by chris-wood** [comment]
I understand that it's editorial, but I'm asking if the new names and types match that which we use in the reference implementation.

**Comment by kevinlewi** [comment]
> I understand that it's editorial, but I'm asking if the new names and types match that which we use in the reference implementation.

Ah I see. Yes, just checked the poc/ directory, and there was no mention of AuthInit/AuthResponse/AuthFinish previously (since we referred to their parameters directly in the ref implementation), so I don't believe anything needs to change there. Thanks for the clarification! :)

**Review by bytemare — APPROVED** [review]

**Comment by kevinlewi** [comment]
@chris-wood -- this should be ready to merge!

**Review by chris-wood — APPROVED** [review]

**Pull request merged**